fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/services/DeleteResourceServiceImpl.java

```diff
@@ -0,0 +1,122 @@
+import org.fcrepo.kernel.api.models.Binary;
+import org.fcrepo.kernel.api.models.Container;
+import org.fcrepo.kernel.api.models.FedoraResource;
+import org.fcrepo.kernel.api.models.ResourceFactory;
+import org.fcrepo.kernel.api.operations.DeleteResourceOperationFactory;
+import org.fcrepo.kernel.api.operations.ResourceOperation;
+import org.fcrepo.kernel.api.services.DeleteResourceService;
+import org.fcrepo.persistence.api.PersistentStorageSession;
+import org.fcrepo.persistence.api.PersistentStorageSessionManager;
+import org.fcrepo.persistence.api.exceptions.PersistentStorageException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class mediates delete operations between the kernel and persistent storage layers
+ *
+ * @author dbernstein
+ */
+public class DeleteResourceServiceImpl implements DeleteResourceService {
+
+    private Logger LOGGER = LoggerFactory.getLogger(DeleteResourceService.class);
+
+    @Inject
+    private ContainmentIndex containmentIndex;
+
+    @Inject
+    private ResourceFactory resourceFactory;
+
+    @Inject
+    private DeleteResourceOperationFactory deleteResourceFactory;
+
+    @Inject
+    private PersistentStorageSessionManager psManager;
+
+    @Override
+    public void perform(final Transaction tx, final FedoraResource fedoraResource) {
+
+        final String fedoraResourceId = fedoraResource.getId();
+
+        try {
+            LOGGER.debug(format("deleting of %s", fedoraResourceId));
+            final PersistentStorageSession pSession = this.psManager.getSession(tx.getId());
+            deleteRecursivelyFromBottom(tx, pSession, fedoraResource);
+        } catch (final PersistentStorageException ex) {
+            throw new RepositoryRuntimeException(format("failed to delete resource %s", fedoraResourceId), ex);
+        }
+    }
+
+
+    private void deleteRecursivelyFromBottom(final Transaction tx, final PersistentStorageSession pSession,
+                                             final FedoraResource fedoraResource) throws PersistentStorageException {
+
+        final String fedoraId = fedoraResource.getId();
+
+        if (fedoraResource instanceof Container) {
+            final Stream<String> children = containmentIndex.getContainedBy(tx, fedoraResource);
+            children.forEach(childResourceId -> {
+                try {
+                    final FedoraResource res = resourceFactory.getResource(tx, childResourceId);
+                    deleteRecursivelyFromBottom(tx, pSession, res);
+                } catch (final PathNotFoundException ex) {
+                    LOGGER.warn(format("Path not found for %s: ignoring.", childResourceId));
+                } catch (final PersistentStorageException ex) {
+                    throw new RepositoryRuntimeException(format("failed to delete resource %s", fedoraId), ex);
+                }
+            });
+        }
+
+        //delete described resource if binary
+        if (fedoraResource instanceof Binary) {
```

Comment:
this should probably be an `else if` of the previous if, since they can't both be true.

Also, I think this method would currently allow the binary description to be deleted independently, which we probably don't want to allow.

Reply:
There is a check for the binary in the FedoraLdp.delete() method.  I could move it here.

Reply:
That would make sense to me